### DISCUSSION TITLE
File Input Redirection for c Command?

### DISCUSSION THREAD

**Discussion by kerszl** [pwndbg/pwndbg#2413] [Ideas]
I use your Pwndbg and I must say it’s much better than Gef, with more options, ranges, etc. But to get to the point, I often use `r < file` in pwndbg to read parameters from a file instead of typing them manually. However, when the program is continued, there’s no option like `c < file`; you have to enter the data manually, which can sometimes be problematic. It is possible to enter data after "continue," but it cannot be read from a file. Is it possible to implement this feature, or is it only available in the original Gdb?

Below example like it looks now:
```
gdb ./chal
GNU gdb (Debian 15.1-1) 15.1
pwndbg> start
pwndbg> b *0x5555555555cd
pwndbg> r < payload1

Breakpoint 2, 0x00005555555555cd in main ()

pwndbg> c < payload2
A syntax error in expression, near `< payload2'.
```

Ps1. The string created by pwndbg (cyclic) is somewhat different from the one created using the pwn cyclic command in the console, and it’s even different in gef... so you have to be careful.

Ps2. It’s great that you fixed the UTF-8 issue; I was getting an error on WSL2:
`Your encoding (utf-8) is different than UTF-8. pwndbg might not work properly. You might try launching GDB with: LC_CTYPE=C.UTF-8 gdb
`

Best regards, and I wish you continued success with the program.

**Comment by disconnect3d** [comment]
Hi,

> (...) or is it only available in the original Gdb?

The `run < inputfile` feature comes from pure GDB. We do not override the `continue` command and I don't think we can handle the `continue < input` thing anyhow. I think it could also be problematic if I think more about it because `run < inputfile` will eventually find EOF in the file -- assuming u read a normal file -- and so this "the file is closed" behavior is something a process can detect and act upon etc. So you can't then just "hey here is another file".

However, what you can do is:
1. Use pwntools and script your thing, and provide different input files anytime etc and use it with gdb via `python script.py GDB LOCAL` (assuming u generate script via `pwn template --host HOST --port PORT ./somebinary > script.py`).
2. Use fifo on Linux. You could do: `mkfifo fifo` and then `run < fifo` and then u can provide more input outside of GDB, via `cat payload1 >> fifo` and then `cat payload2 >> fifo` anytime you want.

> Ps1. The string created by pwndbg (cyclic) is somewhat different from the one created using the pwn cyclic command in the console, and it’s even different in gef... so you have to be careful.

How is that so? We literally use `cyclic` from pwntools: https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/commands/cyclic.py

```
$ pwn cyclic 16 -n 8
aaaaaaaabaaaaaaa
```
vs
```
$ gdb --quiet --batch --ex 'cyclic 16 -n 8'
pwndbg: loaded 165 pwndbg commands and 47 shell commands. Type pwndbg [--shell | --all] [filter] for a list.
pwndbg: created $rebase, $base, $bn_sym, $bn_var, $bn_eval, $ida GDB functions (can be used with print/break)
aaaaaaaabaaaaaaa
```

The only/main difference is that we pass in a default cyclic length based on the architecture of the debugged program. If its 64-bit, we pass in `-n 8` (`--length 8`) whereas the `pwn cyclic ...` command can't know that and does default to `-n 4` (`--length 4`):
```
$ cyclic 16
aaaabaaacaaadaaa
```

> Ps2. It’s great that you fixed the UTF-8 issue

Great to hear that it works fine now :).

**Comment by kerszl** [comment]
Thanx for all answers

Regarding PS1:
I see that now the strings from cyclic are identical, but about half a year ago they were different. Maybe I was checking on pwndbg in a 32-bit application, while pwn cyclic was on 64 bits. But I remember there were some differences, especially at the end, unless they changed something in pwntools? Now it's identical. Thanks for the information.

**Reply by kerszl** [reply]
maybe that was reason?

https://github.com/pwndbg/pwndbg/commit/b5fc45b62e844ddfba382e046c311271f4f0c0c3

**Reply by disconnect3d** [reply]
Probably yeah, nice for checking this out!

**Comment by disconnect3d** [comment]
Closing as I believe this has been resolved :)
